File: Multiple-Way Selection Statements in Dart, Go, JavaScript, Kotlin, PhP, Python, and Rust/Codes/22202103_Esendemir_EmineIrem.dart
void main() {
    int test() {
    return 2;
    }

    print("1. What is the form and type of the control expression?");
    print("1.1 Forms of control expression:");

    // variable
    var x = 2;
    switch (x) {
    case 1:
        print("x is 1");
        break;
    case "2":
        print("x is 2");
        break;
    default:
        print("default case executed");
    }

    // arithmetic expression
    var y = 3;
    switch (x + y) {
    case 4:
        print("x + y is 4");
        break;
    case 5:
        print("x + y is 5");
        break;
    default:
        print("default case executed");
    }

    // literal
    switch ("hello") {
    case "hi":
        print("The string is 'hi'");
        break;
    case "hello":
        print("The string is 'hello'");
        break;
    default:
        print("default case executed");
    }

    // function call
    switch (test()) {
    case 1:
        print("func returned 1");
        break;
    case 2:
        print("func returned 2");
        break;
    default:
        print("default case executed");
    }

    print("1.2 Types of control expression:");

    // int
    int intValue = 3;
    switch (intValue) {
    case 1:
        print("intValue is 1");
        break;
    case 3:
        print("intValue is 3");
        break;
    default:
        print("default case executed");
    }

    // string
    String strValue = "kemal";
    switch (strValue) {
    case "mustafa":
        print("strValue is mustafa");
        break;
    case "kemal":
        print("strValue is kemal");
        break;
    default:
        print("default case executed");
    }

    // double
    double doubleValue = 1.5;
    // Dart 3.0+ is needed for this
    switch (doubleValue) {
    case 1.0:
        print("doubleValue is 1.0");
        break;
    case 1.5:
        print("doubleValue is 1.5");
        break;
    default:
        print("default case executed");
    }

    // Enum
    EnumExample enumValue = EnumExample.B;
    switch (enumValue) {
        case EnumExample.A:
        print("Enum value is A");
        break;
        case EnumExample.B:
        print("Enum value is B");
        break;
        case EnumExample.C:
        print("Enum value is C");
        break;
        case EnumExample.D:
        print("Enum value is D");
        break;
    }

  print("\n\n2. How are the selectable segments specified?");
  var value = 1;
  switch (value) {
    case 1:
      print("case 1");
      var localVar = "inside case 1";
      print("Local variable: $localVar");
      value += 10;
      print("Value after modification: $value");
      
    case 2:
      print("case 2");
      break;
      
    default:
      print("default case executed");
  }

  print("Switch expressions: ");
  x = 2;
  y = switch(x) {
    1 => 2,
    2 => 3,
    _ => 4
  };
  print("x = $x, y = $y");

  print("\n\n3. Is execution flow through the structure restricted to include just a single selectable segment?");

  print("3.1 If no statement exist in the case block, execution will fall through to the next case:");
  var testValue = 1;
  switch (testValue) {
    case 1:
    case 2:  
      print("testValue is either 1 or 2");
    case 3:
      print("testValue is 3");
    default:
      print("default case executed");
  }

  print("3.2 Using continue with label:");
  testValue = 1;
  switch (testValue) {
    case 1:
      print("testValue is 1");
      testValue = 2;
      continue case3; 
    case2: 
    case 2:  
      print("continues to case 2");
      continue case3;
    case3:
    case 3:
      print("continues to case 3");
      break;
    case 4:
      print("not reached");
      break;
    default:
      print("default case executed");
  }

  print("3.3 If a statement exist between cases, it will not fall through:");
  testValue = 1;
  switch (testValue) {
    case 1:
      print("testValue is 1");
    case 2:
      print("testValue is 2");
    case 3:
      print("testValue is 3");
    default:
      print("default case executed");
  }

  print("\n\n4. How are case values specified?");

  // literal
  var name = "kemal";
  switch (name) {
    case "mustafa":
      print("name is mustafa");
      break;
    case "kemal":
      print("name is kemal");
      break;
    default:
      print("default case executed");
  }

  // variable
  var varName = "mustafa"; 
  var secondName = "kemal";
  switch (name) {
    /*case varName:
      print("name is mustafa");
      // Error: Not a constant expression.
      break;*/
    case "kemal":
      print("name is kemal");
      break;
    default:
      print("default case executed");
  }

  // using constants
  const firstName = "kemal";
  switch (name) {
    case firstName:  
      print("name matches firstName constant");
      break;
    default:
      print("default case executed");
  }


  // type patterns
  Object obj = [1, 2, 3];

  switch (obj) {
    case int():
      print("obj is an int");
      break;
    case String():
      print("obj is a String");
      break;
    case List():
      print("obj is a List");
      break;
    default:
      print("obj is something else");
  }

  // list 
  var list = [1, 2, 3];
  switch (list) {
    case [1, 2, 3]:
      print("List contains exactly [1, 2, 3]");
      break;
    case [1, ...]:
      print("List starts with 1");
      break;
    case [..., 3]:
      print("List ends with 3");
      break;
    default:
      print("list is something else");
  }

    print("5. What is done about unrepresented expression values?");

    print("5.1 Using default case:");
    var unrepresented = 4;
    switch (unrepresented) {
    case 1:
        print("number is 1");
        break;
    case 2:
        print("number is 2");
        break;
    case 3:
        print("number is 3");
        break;
    default:
        print("default case executed");
    }


    print("5.2 No default case provided:");
    var A = EnumExample.A;
    switch (A) {
    case EnumExample.A:
        print("Enum value is A");
        break;
    case EnumExample.B:
        print("Enum value is B");
        break;
    // If default is not provided,  Error: The type 'EnumExample' is not exhaustively matched by the switch cases since it doesn't match 'EnumExample.C'.
    default:
        print("default case executed");
    }

    // non-exhaustive switches
    switch (unrepresented) {
    case 1:
        print("number is 1");
        break;
    case 2:
        print("number is 2");
        break;
    case 3:
        print("number is 3");
        break;
    }
}

// Enum declaration for Dart
enum EnumExample {
    A,
    B,
    C,
    D
}

File: Multiple-Way Selection Statements in Dart, Go, JavaScript, Kotlin, PhP, Python, and Rust/Codes/22202103_Esendemir_EmineIrem.go
package main

import "fmt"

func test() int {
	return 2
}
type TestObject struct {
	Name string
}

func main() {
	fmt.Println("1. What is the form and type of the control expression?")
	fmt.Println("\n1.1 Forms of control expression:")

	// variable 
	x := 2
	switch x {
	case 1:
		fmt.Println("x is 1")
	case 2:
		fmt.Println("x is 2")
	default:
		fmt.Println("not found")
	}

	y := 3
	switch x + y {
	case 4:
		fmt.Println("x + y is 4")
	case 5:
		fmt.Println("x + y is 5")
	default:
		fmt.Println("not found")
	}

	// Literals
	switch "a" {
	case "b":
		fmt.Println("The string is 'b'")
	case "a":
		fmt.Println("The string is 'a'")
	default:
		fmt.Println("not found")
	}

	// function call
	switch test() {
	case 1:
		fmt.Println("func returned 1")
	case 2:
		fmt.Println("func returned 2")
	default:
		fmt.Println("not found")
	}

	fmt.Println("\n1.2 Types of control expression:")
	// integer type
	intValue := 3
	switch intValue {
	case 1:
		fmt.Println("intValue is 1")
	case 3:
		fmt.Println("intValue is 3")
	}

	// string type
	strValue := "a"
	switch strValue {
	case "a":
		fmt.Println("strValue is 'a'")
	case "b":
		fmt.Println("strValue is 'b'")
	}

	// boolean type
	boolValue := true
	switch boolValue {
	case true:
		fmt.Println("it is true")
	case false:
		fmt.Println("it is false")
	}

	// struct type
	person1 := TestObject{Name: "Emine"}
	person2 := TestObject{Name: "İrem"}

	switch person1 {
	case person2:
		fmt.Println("Matched person2")
	case TestObject{Name: "Emine"}:
		fmt.Println("Matched Emine")
	default:
		fmt.Println("not found")
	}

	// tagless switch 
	x = 1
	switch {
	case x < 0:
		fmt.Println("x is negative")
	case x == 0:
		fmt.Println("x is zero")
	case x > 0:
		fmt.Println("x is positive")
	}

	// type switch
	var typeSwitch1 interface{} = "a"
	switch typeSwitch1.(type) {
	case int:
		fmt.Println("typeSwitch1 is int")
	case string:
		fmt.Println("typeSwitch1 is string")
	case bool:
		fmt.Println("typeSwitch1 is bool")
	case TestObject:
		fmt.Println("typeSwitch1 is TestObject")
	default:
		fmt.Println("typeSwitch1 is an unknown type")
	}

	fmt.Println("\n\n2. How are the selectable segments specified?")

	value := 2
	switch value {
	case 1:
		fmt.Println("1 is executed")
		value *= 2
		fmt.Println("Value doubled:", value)
	case 2, 3, 4: 
		fmt.Println("2, 3 or 4 is executed")
		local := "inside case 1"
		fmt.Println("Local variable:", local)
		fmt.Println("Function result:", test())
		value += 10
		fmt.Println("Value after modification:", value)
	default:
		fmt.Println("Default case")
	}

	fmt.Println("\n\n 3. Is execution flow through the structure restricted to include just a single selectable segment?")
	fmt.Println("3.1 no fallthrough):")
	testValue := 1
	switch testValue {
	case 1:
		fmt.Println("testValue is 1")
		testValue = 2
	case 2:  
		fmt.Println("testValue is 2")
	/*case 1:
		fmt.Println("Case 1 matched again")
	Error: duplicate case 1 (constant of type int) in expression switch
	*/
	}
	fmt.Println("new testValue:", testValue)

	fmt.Println("3.2 Using fallthrough:")
	testValue = 1
	switch testValue {
	case 1:
		fmt.Println("Case 1 exectuted")
		fallthrough  
	case 2:  
		fmt.Println("Execution continues to case 2")
		fallthrough
	case 3:
		fmt.Println("Execution continues to case 3")
	case 4:
		fmt.Println("This won't execute")
	}
	fmt.Println("testValue after fallthrough switch:", testValue)

	fmt.Println("\n4. How are the case values specified?")

	// literal
	name := "irem"
	switch name {
	case "emine":
		fmt.Println("name is emine")
	case "irem":
		fmt.Println("name is irem")
	}

	/*wrongtype := "irem"
	switch wrongtype {
	case 1:
		fmt.Println("wrongtype is 1")
	default:
		fmt.Println("wrongtype is not 1")
	}
	// Error: cannot convert 1 (untyped int constant) to type string
	*/

	// variable
	trialCase := "irem"
	switch name {
	case trialCase: 
		fmt.Println("trialCase is same as name")
	case "irem":
		fmt.Println("name is irem")
	}

	// expression
	num := 5
	num1 := 3
	num2 := 2
	switch num {
	case num1 + num2:  
		fmt.Println("num is equal to num1 + num2")
	case 10 - 2:
		fmt.Println("num is equal to 10 - 2")
	}

	// multiple values
	multipleValueNum := 1
	multipleValueNum1 := 1
	multipleValueNum2 := 2
	multipleValueNum3 := 3
	switch multipleValueNum {
	case multipleValueNum1, multipleValueNum2:
		fmt.Println("the number is one of these two")
	case multipleValueNum3:
		fmt.Println("the number is 3")
	}

	// tagless switch
	taglessValue := 1
	switch {
	case taglessValue < 0:
		fmt.Println("taglessValue is negative")
	case taglessValue == 0:
		fmt.Println("taglessValue is zero")
	case taglessValue > 0:
		fmt.Println("taglessValue is positive")
	}

	// type switch
	var i interface{} = "irem"
	switch i.(type) {
	case int:
		fmt.Println("i is of type int")
	case string:
		fmt.Println("i is of type string")
	case bool:
		fmt.Println("i is of type bool")
	case TestObject:
		fmt.Println("i is of type TestObject")
	default:
		fmt.Println("i is of an unknown type")
	}
	fmt.Println("\n5.What is done about unrepresented expression values?");
	fmt.Println("5.1 Using default case:")
	defaultInt := 3
	switch defaultInt {
	case 1:
		fmt.Println("defaultInt is 1")
	case 2:
		fmt.Println("defaultInt is 2")
	default:
		fmt.Println("defaultInt is not 1 or 2")
	}

	fmt.Println("5.2 No default case provided:")
	switch defaultInt {
	case 1:
		fmt.Println("defaultInt is 1")
	case 2:
		fmt.Println("defaultInt is 2")
	}	
}